### DISCUSSION TITLE
Add support for the declension of the names of months?

### DISCUSSION THREAD

**Discussion by ryszardpacer** [datenstrom/community#578] [Ask a question] [answered]
I am impressed with Yellow's simplicity and elegance. I would like to use it in my new blog. 
The only obstacle I see is the imprecise display of the date in Polish. The declination of the names of the months is absolutely necessary. IMO ;-) While in English the names of months always have this nominative form, in Polish there must be a genitive in the date. 
Instead of 12 grudzień (December 12), it should be 12 grudnia (December 12). Just like on my website: pacer(dot)pl/teksty I used the strftime formula here: 
`%Y %m %d `
However, in Yellow I cannot implement it. Maybe you can give me a clue? Or maybe it could be included in the Polish translation of Yellow? It would be perfect. 

I am not an IT specialist and I speak little English, so please be understanding if my problem seems trivial.

Rysiek

**Comment by schulle4u** [comment]
Maybe @kanbeq can give feedback on this topic?

**Comment by kanbeq** [comment]
The translation file is set like this:
`CoreDateMonths: styczeń, luty, marzec, kwiecień, maj, czerwiec, lipiec, sierpień, wrzesień, październik, listopad, grudzień`
Please try to replace the above translation line with the following one and check if it works properly:
`CoreDateMonths: stycznia, lutego, marca, kwietnia, maja, czerwca, lipca, sierpnia, września, października, listopada, grudnia`

I am not sure if I can introduce this to the general translation, because it was Mr. Rysiek who took this form of dating. (Because he start dates with the word "Opublikowano"). Even in English, if you want to start with the past form of the verb ("Published"), you would have to add endings to each day: st, rd, th, etc. I don't know English perfectly, so I don't know if it should be like that. But I think so.

I have a numerical date on my website, which is enough for me. I also believe that readers will not be privy to how the date is displayed, but rather to the content of the post (but this is my general thought).

**Reply by ryszardpacer** [reply]
It works. Thank you very much @kanbeq  and @schulle4u  for your help. It was that simple.

It is true that the content is essential. I am probably a little skewed about formal correctness.

**Reply by schulle4u** [reply]
Please note that updates to the language extension files can overwrite your modifications. I suggest to put them in `system/extensions/yellow-language.ini` instead. See also: https://datenstrom.se/yellow/help/adjusting-system#language-settings

**Reply by markseuffert** [reply]
I wonder if we can find a permanent solution. At the moment we're using `CoreDateMonths` for the language specific month names. Can you be so kind and clarify the following, do you use both the nominative and the genitive form for dates in the Polish language? For example is there a difference between "15 December 2020" and "December 2020" in the Polish language?

**Reply by kanbeq** [reply]
It looks like I wrote it wrong before. When giving dates in Polish, only the genitive form is used. So we have a permanent solution after all.

By the way, if we are already on the topic of dates, there is still a problem with adding translations to months: one month ago, two months ago, three months ago, four months ago, five months ago, etc. (what in Polish looks like that: jeden miesi**ąc** temu, dwa miesi**ące** temu, trzy miesi**ące** temu, cztery miesi**ące** temu, pięć miesi**ęcy** temu). It is similar with the years and the future form. In the translation file `CoreDatePast` and `CoreDateFuture` are responsible for this. I skipped this problem by using the parenthesized variant of these words, but it doesn't look very nice and isn't the best solution.

**Reply by markseuffert** [reply]
Okay please send a pull request with the genitive form, if you have time. Just in case this topic comes up again, perhaps someone wants to make a calendar extension or a language (like Finnish?) differentiates grammatical cases, I wouldn't mind having `CoreDateMonthsNominative` and `CoreDateMonthsGenitive` in the language settings.

Can you elaborate for non-polish speakers, what are the rules for relative dates?

**Reply by ryszardpacer** [reply]
I try it ;-)

Brief information about the inflection of the names of the months, the words month and the words year in Polish.

In dates, the month should be in the genitive, as it was already said: nominative - **styczeń** (January), genitive -  **1 stycznia 2020 r.** (January 1, 2020). Add the word year or **r. (roku)** for short is elegant. This is due to the fact that we think about the day. We do not write the word day, because we do not like to repeat ourselves, but the number tells us that is the number of the day of the month. This changes the form of the month noun.

As kanbeq mentioned, the inflection of the word month is more complex. We use the nominative singular **miesiąc** (month - we omit the word "one" as obvious) and plural **2 miesiące, 3 miesiące, 4 miesiące** (two months, three months, four months).

It is not known why, one of these grammar puzzles, from 5 upwards we use the genitive **5 miesięcy , 6 miesięcy, 7 miesięcy 8 miesięcy, 9, 10, 11… 180 miesięcy** (5 months, 6 months, 7 months, 8 months, 9, 10, 11 ... 180 months etc.).

Of course, we are only talking about distance in time. For example: Zdarzyło się to **dwa miesiące** temu (It happened two months ago). However, if it were for an action and were to describe the passage of time in terms of other actions, we would have to use the genitive also for the numbers 1 through 4. For example: Brakło mu **miesiąca** (He was short of a month or He was missing a month.). Brakło mu **2 miesięcy** (He lacked 2 months). Or any other of the grammatical case on the context.

The word **rok** (year) is declined in the same way as **miesiąc** (month). **Jeden rok** (one year or just a year) is the nominative. **2 lata** (two years) plural nominative up to 4 years, analogous to the word month. From the number five upwards, we are **5 lat, 6 lat…** (5, 6…). The only difference is in the root of the word. The singular is the rok (year) form, the plural is lata (years),  **rok, 2 lata, 3 lata, 4 lata, 5, 6… lat** (1 year, 2, 3, 4 years, 5, 6 ... years). **Lata** (years) is the nominative, **lat** (years) is the genitive.

**Reply by markseuffert** [reply]
Thank you for the detailed explanation. 

I added support for genitive month names, but I left the algorithm for the relative dates untouched. You should now be able to configure date formats more flexible. You can use PHP's [date format](https://www.php.net/manual/en/datetime.format.php). Where  `F` is the nominative form of a month, for example **grudzień**,  and `V` is the genitive form of a month, for example **grudnia**. This means there's no automatic detection which form to use. You have to choose it manually.

It looks like there are more languages with genitive month names:
https://en.wikipedia.org/wiki/Date_format_by_country  
https://rluzynski.fedorapeople.org/slides/2017-01-27-DevConf.cz/GenitiveMonths-updated.pdf

**Comment by markseuffert** [comment, marked as answer]
You can configure the [date format](https://www.php.net/manual/en/datetime.format.php) in file `system/extensions/yellow-language.ini`. Here are some examples for Polish:

`CoreDateFormatMedium: Y-m-d` = 2020-12-06
`CoreDateFormatMedium: d-m-Y` = 06-12-2020
`CoreDateFormatMedium: j F Y \r` = 6 grudzień 2020 r
`CoreDateFormatMedium: j V Y \r` = 6 grudnia 2020 r

I tested this with a new installation,  you might have to update your website to the latest version.

**Reply by ryszardpacer** [reply]
Works great. Thank you very much!
